Image spec example

pingsutw opened the pull request:
Add an example for using imageSpec in the task

preview: https://flyte--988.org.readthedocs.build/projects/cookbook/en/988/auto/core/image_spec/image_spec.html

<img width="1873" alt="image" src="https://user-images.githubusercontent.com/37936015/236450652-82e196b9-5075-45ae-871f-3839d56d7919.png">

pingsutw commented:
@zeryx @peridotml @cosmicBboy mind taking a look

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+"""
+.. _image_spec_example:
+
+Building Image without Dockerfile
```
```suggestion
Building Docker Images without a Dockerfile
```

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+"""
```
Can you fix the name of the file: sepc to spec?

pingsutw replied:
oops, thanks

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+you can specify rules for binding container images. By default, flytekit binds a single container image, i.e.,
+the `default Docker image <https://ghcr.io/flyteorg/flytekit>`__, to all tasks. To modify this behavior,
+use the ``container_image`` parameter available in the :py:func:`flytekit.task` decorator, and pass an
+`ImageSpec`.
```
```suggestion
``ImageSpec``.
```

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+#    :class: important
+#
+#    - Install `flytekitplugins-envd <https://github.com/flyteorg/flytekit/tree/master/plugins/flytekit-envd>`__ to build the image spec.
+#    - To build the image on remote machine, check this `doc <https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine>`__
```
```suggestion
#    - To build the image on remote machine, check this `doc <https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine>`__.
```

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+#    - Install `flytekitplugins-envd <https://github.com/flyteorg/flytekit/tree/master/plugins/flytekit-envd>`__ to build the image spec.
+#    - To build the image on remote machine, check this `doc <https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine>`__
+#
+#
```
Let's begin by importing the necessary libraries.

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+)
+
+# %%
+# ``is_container`` is used to check if the task is using the image built from the image spec.
+# If the task is using the image built from the image spec, then the task will import tensorflow.
+# It can reduce module loading time and avoid unnecessary dependency installation in the single image.
```
```suggestion
# ``is_container`` is used to determine whether the task is utilizing the image constructed from the ``ImageSpec``.
# If the task is indeed using the image built from the ``ImageSpec``, it will then import Tensorflow.
# This approach helps minimize module loading time and prevents unnecessary dependency installation within a single image.
```

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+
+
+# %%
+# Both ``get_pandas_dataframe`` and ``train_model`` will use the image built from the image spec.
```
Can you move `train_model` to this section?

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+
+
+# %%
+# Users can also pass imageSpec yaml file to the ``pyflyte run`` or ``pyflyte register`` command to override the
+# container_image. For instance:
```
```suggestion
# There exists an option to override the container image by providing an Image Spec YAML file to the ``pyflyte run`` or ``pyflyte register`` command.
# This allows for greater flexibility in specifying a custom container image. For example:
```

samhita-alla commented on cookbook/core/containerization/image_sepc_example.py:
```diff
@@ -0,0 +1,117 @@
+
+
+# %%
+# Both ``get_pandas_dataframe`` and ``train_model`` will use the image built from the image spec.
```
If you move all the code into one section (tasks + workflow), you can say "To enable tasks to utilize the images built with ``ImageSpec``, you can specify the ``container_image`` parameter for those tasks.".

wild-endeavor commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,116 @@
+# These specified packages will be added on top of the `default image <https://github.com/flyteorg/flytekit/blob/master/Dockerfile>`__, which can be found in the Flytekit Dockerfile.
+# If desired, you can also override the default image by providing a custom ``base_image`` parameter when using the ``ImageSpec``.
+pandas_image_spec = ImageSpec(
+    base_image="ghcr.io/flyteorg/flytekit:py3.8-1.6.0",
```
should we mention `DefaultImages.default_image()` also?

pingsutw replied:
Added it.

samhita-alla commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,118 @@
+    return LogisticRegression(max_iter=max_iter, multi_class=multi_class)
+
+
+# Get a basic model to train.
```
This isn't required.

samhita-alla commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,118 @@
+    return model
+
+
+# A simple pipeline to train a model.
```
```suggestion
# Lastly, let's define a workflow to capture the dependencies between the tasks.
```

samhita-alla commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,118 @@
+
+
+# %%
+# To enable tasks to utilize the images built with ImageSpec, you can specify the container_image parameter for those tasks.
```
```suggestion
# To enable tasks to utilize the images built with ``ImageSpec``, you can specify the ``container_image`` parameter for those tasks.
```

samhita-alla commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,118 @@
+# %%
+# You can specify python packages, apt packages, and environment variables in the ``ImageSpec``.
+# These specified packages will be added on top of the `default image <https://github.com/flyteorg/flytekit/blob/master/Dockerfile>`__, which can be found in the Flytekit Dockerfile.
+# More specifically, flytekit will call `DefaultImages.default_image() <https://github.com/flyteorg/flytekit/blob/f2cfef0ec098d4ae8f042ab915b0b30d524092c6/flytekit/configuration/default_images.py#L26-L27>`__, this function will
+# return the default image based on the python version and flytekit version. For example, if you are using python 3.8 and flytekit 0.16.0, the default image will be ``ghcr.io/flyteorg/flytekit:py3.8-1.6.0``.
```
```suggestion
# More specifically, flytekit invokes `DefaultImages.default_image() <https://github.com/flyteorg/flytekit/blob/f2cfef0ec098d4ae8f042ab915b0b30d524092c6/flytekit/configuration/default_images.py#L26-L27>`__ function.
# This function determines and returns the default image based on the Python version and flytekit version. For example, if you are using python 3.8 and flytekit 0.16.0, the default image assigned will be ``ghcr.io/flyteorg/flytekit:py3.8-1.6.0``.
```

samhita-alla commented on cookbook/core/containerization/image_spec_example.py:
```diff
@@ -0,0 +1,118 @@
+#    :class: important
+#
+#    - Install `flytekitplugins-envd <https://github.com/flyteorg/flytekit/tree/master/plugins/flytekit-envd>`__ to build the image spec.
+#    - To build the image on remote machine, check this `doc <https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine>`__
```
```suggestion
#    - To build the image on remote machine, check this `doc <https://envd.tensorchord.ai/teams/context.html#start-remote-buildkitd-on-builder-machine>`__.
```

jboutwell-rva commented:
I just recently tried out the ImageSpec feature and its great, but there are a few things that were not obvious to a new envd user that might be worth mentioning in the help:
- registry="" by default which is useful for local development, but in practice a registry is required for Flyte to run remotely.  The example given using registry="pingsutw" works because there is a Docker Hub account with username "pingsutw" (see https://hub.docker.com/r/pingsutw/flytekit).  A free account is sufficient for testing.  The registry is not limited to Docker Hub, an example for AWS ECR is registry="12345.dkr.ecr.us-east-1.amazonaws.com/mygroup".  In this case the repository "12345.dkr.ecr.us-east-1.amazonaws.com/mygroup/flytekit" must first exist.  "flytekit" is the default name set here https://github.com/flyteorg/flytekit/blob/d7060cc13be711ea37773435abcc7fbca6c3e4df/flytekit/image_spec/image_spec.py#L38
- when using a registry `docker login ...` is required to push the image

pingsutw commented:
running an issue, envd can't build an image inside docker container. investigating

pingsutw commented:
@samhita-alla Thanks for the review and help. I've fixed the test errors.

samhita-alla commented on cookbook/core/image_spec/image_spec.py:
```diff
@@ -0,0 +1,118 @@
+
+sklearn_image_spec = ImageSpec(
+    base_image="ghcr.io/flyteorg/flytekit:py3.8-1.6.0",
+    packages=["tensorflow"],
```
```suggestion
    packages=["scikit-learn"],
```

cosmicBboy commented:
@pingsutw @samhita-alla is there a particular reason we created a new "ImageSpec" section instead of putting it inside "Containerization"? 
![image](https://github.com/flyteorg/flytesnacks/assets/2816689/e332b3a6-3bee-46c1-a11d-4db3c4240ebd)

pingsutw commented:
I moved to a new session since the GA workflow restricts us from pushing the image (flytekit will build and push the image for ImageSpec) in the pull request. Consequently, I have to create a new folder with a makefile that enables us to bypass the serialization of the workflow in the CI process.

I also plan to add more docs for imageSpec, like remote build, envd private registry, so I added a new session.

cosmicBboy commented:
gotcha, ok cool

cosmicBboy commented:
the guide should probably include:
- installing `pip install flytekitplugins-envd` to make this work with envd
- how to build and push an image with `pyflyte build`

samhita-alla commented:
Can we merge this?

pingsutw commented:
@samhita-alla yes, we can merge this pr
